### DISCUSSION TITLE
Harsher penalties for low effort duplicate ride types?

### DISCUSSION THREAD

**Discussion by shanemcguffin** [OpenRCT2/OpenRCT2#17042] [Feature requests]
```
// Add guest score for ride type
suggestedMaxGuests += ride.GetRideTypeDescriptor().BonusValue;
```

**You can spam 20 Launched Free Falls for a 1300 suggested guest maximum.** This seems broken as it is an easy way to beat most scenarios and it discourages creativity and creating beautiful parks.

What if we do something like...

```
// Add guest score for ride type
for (auto rideType : GetAllRideTypes()) {
    for (int i = 1; i <= GetRidesOfType(rideType).size(); i++) {
        if (towerRide || !trackRide) {
            suggestedMaxGuests += ride.GetRideTypeDescriptor().BonusValue / i;
        } else {
            // rideRatingNormalized is normalized percentage between 0 and 1.
            suggestedMaxGuests += ride.GetRideTypeDescriptor().BonusValue * rideRatingNormalized;
        }
    }
}
```

**This penalizes guest max for low effort track rides.**

New suggested max guest counts per number of Launched Free Falls:

1. 65
2. 97
3. 118
4. 134
5. 147
6. 157
7. 166
8. 174
9. 181
10. 187
11. 192
12. 197
13. 202
14. 206
15. 210
16. 214
17. 217
18. 220
19. 223
20. 226

Now those **20 Launched Free Falls get a suggested max guest count of 226 instead of 1300.** Having 3 of them still gets a max bump of 118, which seem fair because maybe you just want to have a few of them scattered throughout your park.

You still get *some* value for the first few duplicates, but it quickly falls off. Maybe add a cheat for players who want to keep the old behavior.

We could even **go a step further to scale the ride value penalty of duplicate ride types.**

```
// ride_ratings_calculate_value()
...
if (otherRidesOfSameType > 1)
        value -= value / 4;
```
Instead, what if we said...

```
if (trackRide && !towerRide && otherRidesOfSameType > 1) {
    value -= value / 4;
}
// dampens non-track / tower ride value by up to 50% depending on number of otherRidesOfSameType
// start with 75% value, then 70%, 65%, etc.
else if (otherRidesOfSameType > 1) {
    float percentDampener = (80.0 - 5 * otherRidesOfSameType) * .01;
    value = max(value * percentDampener, value * .50);
}
```

**Comment by ocalhoun6** [comment]
If you're going to do this, you should probably do something about coasters and other tracked rides as well, addressing Marcel's favorite exploit of the tiny launched corkscrew.

Since large custom coasters of the same type can be very different from each other, though, maybe its best to only apply the extra duplicate ride penalty if the coaster is under a certain length?

-----

Of course, *ideally*, we'd have the game actually check tracked rides to see if they're identical in layout to other tracked rides in the park. Penalizing them heavily like this if they're the same, while giving them little to no penalty if they have different track layouts. But maybe that would be too difficult to code and to expensive in system resources?

One shortcut might be to have the game look for other rides of the same ride type and equal track length. That would be fairly easy to do, and the only downside is that it might _rarely_ have a false positive if two coasters randomly happen to be exactly the same length. Perhaps it could be combined with another statistic such as number of drops. It would be _extremely_ rare for two coasters to have the same length and same number of drops without being duplicate designs (unless somebody is deliberately _trying_ to do that).

So it would be like:
```
if (there are other rides of the same type in the park already)
    apply the normal duplicate ride penalty
    if (it's a tracked ride type && ride length == any of the other rides of the same type)
        if (number of drops == the possible duplicate)
            apply the additional penalty
```

**Reply by James103** [reply]
Parkitect has a mechanism that causes rides to lose a percentage of their stats and value if the ride time is too short. Could that be applied here to OpenRCT2 as well? That way, most actual coasters would get their normal stats as long as their ride time is at least 30-45 seconds, but microcoasters would get progressively lower stats because of their short runtime.

Quoting from the [Parkitect wiki](https://parkitect.fandom.com/wiki/Coaster_Stat_Calculation_Guide#Excitement_Penalties): 
> Additionally, if a coaster has a duration of less than 35 seconds, it loses a certain percentage of its excitement rating, up to 75% of its excitement rating if its duration is 0 seconds.

**Reply by ocalhoun6** [reply]
RCT already has stat penalties on _some_ coasters if the ride length is too short. But it's on a per-coaster basis, and some coasters have no minimum length requirement.

That makes sense somewhat, because some coaster types can build shuttle loop type designs, and I don't think your standard shuttle loop should be penalized for being short. _But_ maybe there _should_ be a very, very short minimum length limit -- something that a basic shuttle loop like design would satisfy, but a two-tile long launched corkscrew would not satisfy.

**Reply by James103** [reply]
At the very least, the following three stat requirements should be added:
* The Air Powered Vertical Coaster should have at least one drop.
* The Corkscrew Roller Coaster should have a minimum length of at least 50 m and a top speed of at least 42 km/h.

Doing this should nerf the overpowered designs shown by Marcel Vos by a factor of 2 for the Air Powered Vertical Coaster and a factor of 4 for the Corkscrew Roller Coaster.

If other microcoaster designs are to be nerfed, a universal minimum length requirement of 50 m could be added to all tracked ride types, not just the Corkscrew Roller Coaster.

**Comment by Die4Ever** [comment]
if you really wanna check for identical rides then you could create hash values for each ride, just a hash of all the track pieces in it

even something simple like a crc32

but I'm not sure this needs to be fixed at all? I'm not opposed to fixing it, but if players want to cheat then they can just use the cheat menu, using this exploit is basically the same thing where it's a consensual "cheat", it's not something that you exploit by accident

it might be better to just fix it with a couple additional stat requirements like James103 said

**Comment by toimine** [comment]
Rct2 always stroke me as a game that rewarded various play styles, where you could do anything you'd like as long as your park thrives, adding extra penalties to direct the player in a certain direction feels wrong

**Comment by ageekhere** [comment]
What could be an idea is with roller coasters that are of the same type check how similar they are to each other.
![ride](https://user-images.githubusercontent.com/13173133/215363390-9e85a035-1e15-4ba0-a3d1-db0f7e645783.png)

If a patten is detected then check how similar the ride is. If the roller coasters similarity is > then x % then a penalty is given. 

The method that could be used is what they use for plagiarism checkers, look for matching pattens are see how similar they are.

To give feedback to the player peeps could say things like " this ride is too similar to other rides in the park and I want to go on something new"
